src/errors.ts

```diff
@@ -1,132 +1,266 @@
-import type { IncomingHttpHeaders } from "http"
-import type {
-  HTTPError as GotHTTPError,
-  TimeoutError as GotTimeoutError,
-  Response as GotResponse,
-} from "got"
+import { SupportedResponse } from "./fetch-types"
 import { isObject } from "./helpers"
+import { Assert } from "./type-utils"
 
-export class RequestTimeoutError extends Error {
-  readonly code = "notionhq_client_request_timeout"
+/**
+ * Error codes returned in responses from the API.
+ */
+export enum APIErrorCode {
+  Unauthorized = "unauthorized",
+  RestrictedResource = "restricted_resource",
+  ObjectNotFound = "object_not_found",
+  RateLimited = "rate_limited",
+  InvalidJSON = "invalid_json",
+  InvalidRequestURL = "invalid_request_url",
+  InvalidRequest = "invalid_request",
+  ValidationError = "validation_error",
+  ConflictError = "conflict_error",
+  InternalServerError = "internal_server_error",
+  ServiceUnavailable = "service_unavailable",
+}
+
+/**
+ * Error codes generated for client errors.
+ */
+export enum ClientErrorCode {
+  RequestTimeout = "notionhq_client_request_timeout",
+  ResponseError = "notionhq_client_response_error",
+}
+
+/**
+ * Error codes on errors thrown by the `Client`.
+ */
+export type NotionErrorCode = APIErrorCode | ClientErrorCode
+
+/**
+ * Base error type.
+ */
+abstract class NotionClientErrorBase<
+  Code extends NotionErrorCode
+> extends Error {
+  abstract code: Code
+}
+
+/**
+ * Error type that encompasses all the kinds of errors that the Notion client will throw.
+ */
+export type NotionClientError =
+  | RequestTimeoutError
+  | UnknownHTTPResponseError
+  | APIResponseError
+
+// Assert that NotionClientError's `code` property is a narrow type.
+// This prevents us from accidentally regressing to `string`-typed name field.
+type _assertCodeIsNarrow = Assert<NotionErrorCode, NotionClientError["code"]>
+
+// Assert that the type of `name` in NotionErrorCode is a narrow type.
+// This prevents us from accidentally regressing to `string`-typed name field.
+type _assertNameIsNarrow = Assert<
+  "RequestTimeoutError" | "UnknownHTTPResponseError" | "APIResponseError",
+  NotionClientError["name"]
+>
+
+/**
+ * @param error any value, usually a caught error.
+ * @returns `true` if error is a `NotionClientError`.
+ */
+export function isNotionClientError(
+  error: unknown
+): error is NotionClientError {
+  return isObject(error) && error instanceof NotionClientErrorBase
+}
+
+/**
+ * Narrows down the types of a NotionClientError.
+ * @param error any value, usually a caught error.
+ * @param codes an object mapping from possible error codes to `true`
+ * @returns `true` if error is a `NotionClientError` with a code in `codes`.
+ */
+export function isNotionClientErrorWithCode<Code extends NotionErrorCode>(
```

Comment:
I think we should only export `isNotionClientError` for now. The error type can be narrowed with equality checks and passing in an object map is cumbersome.

```suggestion
function isNotionClientErrorWithCode<Code extends NotionErrorCode>(
```